Adding Support for OpenSSL 3.x and Lazarus

MWASoftware opened the pull request:
This pull request provides a full update ot the package. It is based on the OpenSSL headers generated by the code generator provided by https://github.com/MWASoftware/PascalAPI4OpenSSL

This version supports OpenSSL 3.x, 1.1.1 and 1.0.2 for dynamic library loading. It also supports both Lazarus/fpc and Delphi. Package build files are provided for each toolchain.

Three link models are supported.
   * Dynamic Library Load (the default and the approach used in previous versions)
   * compile time linkage to a shared (.so or .dll) library (OpenSSL 3.x only)
   * compile time linkage to a static library (FPC only with gcc compiled OpenSSL).

For dynamic library load, a "Just in Time" approach is used where each API call is initialised to a local proc "loader" function.
The intent is that on the first call a given API function, the actual entry point in the OpenSSL function
is loaded and the API call is set to the loaded entry point. The API function is now called on the user's
behalf. If the call fails to load then it is replaced by a compatibility function (if one exists). If none exists 
then an exception is raised. If an API function is allowed to be nil (as set in the template), then the function
is loaded at library load time.

See the README file for further information.

MWASoftware committed:
Initial reoriganisation of subdirs with files split into src and delphi subdirs for pascal source and delphi build files respectively.

MWASoftware committed:
Updated source tree from OpenSSLFinal branch of Indy.ProposedUpdate

MWASoftware committed:
Added Tests and Lazarus Packages

MWASoftware committed:
Updated Delphi Project Files

MWASoftware committed:
Removed test project group

MWASoftware committed:
Updated README to include OpenSSL info

MWASoftware committed:
Fixed reference to companion dpk

MWASoftware committed:
Updated references to source files in Delphi project file

MWASoftware committed:
Updated to README

MWASoftware committed:
Added a gitattributes file

MWASoftware committed:
Copy fix for Indy.ProposedUpdate Issue#10 to this package

MWASoftware committed:
Updated package from Indy.ProposedUpdate August 2025 Update in order to include updated unit files

MWASoftware committed:
Updated Test Programs

MWASoftware committed:
Update dependencies in indyopenssl fpc package

MWASoftware committed:
Updated Delphi Packages

RoLex commented:
Hi.

How to correctly install this package with main Indy package from git? Delphi 12.

When I try to build this package "as is" after installing Indy package from git (successfully), first errors that appear is missing RSOSSLEVPDigestExError and RSOSSLEVPDigestUpdateError from IdResourceStringsOpenSSL.pas. This is because IndyCore package already contains IdResourceStringsOpenSSL.pas without these constants. Next comes the MD5 and SHA errors, function declarations differ from previous ones. I assume same thing there, those functions were already declared earlier.

What I'm actually having trouble to understand, is how to merge https://github.com/IndySockets/Indy and https://github.com/MWASoftware/IndyTLS-OpenSSL in order to test them.

Regards.

MWASoftware commented:
If you are trying to test out my proposed OpenSSL add on for Indy, then you must use the IndySockets  Indy10.7 branch and not the master branch. Please check that you are on the correct branch.

> Hi.
> 
> How to correctly install this package with main Indy package from git? Delphi 12.
> 
> When I try to build this package "as is" after installing Indy package from git (successfully), first errors that appear is missing RSOSSLEVPDigestExError and RSOSSLEVPDigestUpdateError from IdResourceStringsOpenSSL.pas. This is because IndyCore package already contains IdResourceStringsOpenSSL.pas without these constants. Next comes the MD5 and SHA errors, function declarations differ from previous ones. I assume same thing there, those functions were already declared earlier.
> 
> What I'm actually having trouble to understand, is how to merge https://github.com/IndySockets/Indy and https://github.com/MWASoftware/IndyTLS-OpenSSL in order to test them.
> 
> Regards.

RoLex commented:
Ouch, missed that one. Will do, tyvm!

MWASoftware committed:
Include readme on IdSSLOpenSLL split

RoLex commented:
I think I got it with Indy-10.7 branch, but not sure about Linux..

When I add IdSSLOpenSSL in uses in my project, compiler tries to compile this:

```
{$IFDEF WINDOWS}
Windows,
{$ENDIF}
```

Meaning that Windows is defined, while chosen platform is Linux.

What have I missed?

RoLex commented:
`[dcc32 Fatal Error] E2202 Required package 'IndyOpenSSL290' not found`

Got this when trying to compile your test project `openssl-server` under Windows.

I think there is a miss with package naming somewhere. Is it IndyOpenSSL290 or IndyTLSOpenSSL290? :-D

MWASoftware commented:
Ah, you are obviously using Delphi.

There is a naming mismatch between Lazarus/fpc and Delphi which I failed to spot. I predominately use the former, but I do try to also repeat the tests with Delphi. Looks like I missed this one. 

I will update and ensure consistent package names. In the meantime, IndyTLSOpenSSL290 is the correct package name for Delphi.

RoLex commented:
Great, thanks!

What about this snippet that is being compiled under Linux platform:

```
{$IFDEF WINDOWS}
Windows,
{$ENDIF}
```

Do you have any idea?

Regards.

MWASoftware commented:
> I think I got it with Indy-10.7 branch, but not sure about Linux..
> 
> When I add IdSSLOpenSSL in uses in my project, compiler tries to compile this:
> 
> ```
> {$IFDEF WINDOWS}
> Windows,
> {$ENDIF}
> ```
> 
> Meaning that Windows is defined, while chosen platform is Linux.
> 
> What have I missed?

That is surprising  WINDOWS is defined in the IdCompilerDefines.inc file that is standard across Indy. My Linux testing is with Lazarus/fpc and I have not seen this problem. What is your Linux build environment?

RoLex commented:
I use Ubuntu 24.04 x64 as SDK imported with PAServer 23.0 in Delphi 12.

```
<MyProject.dpr>

// before IdCompilerDefines.inc is included

{$IFNDEF FPC}
  // is defined
{$ENDIF}

{$IFDEF WINDOWS}
  // not defined
{$ENDIF}

{$IFDEF MSWINDOWS}
  // not defined
{$ENDIF}
```

```
<IdSSLOpenSSL.pas>

// after IdCompilerDefines.inc is included

{$IFDEF WINDOWS}
Windows, // is defined
{$ENDIF}
```

You are right, IdCompilerDefines.inc defines WINDOWS if compiler is _not_ FPC. So basicaly need to add more defines related to chosen build platform. I'll play with this file, thank you!

rlebeau commented:
> You are right, IdCompilerDefines.inc defines WINDOWS if compiler is _not_ FPC.

Yes, but only if `MSWINDOWS` is defined by the compiler.  However, it does also manually define `MSWINDOWS` prior to Delphi 6, so is it possible that `VCL_6_OR_ABOVE` is somehow not being defined correctly for your Delphi compiler?

RoLex commented:
Same with `IdOpenSSLHeaders_crypto.pas`, line: `{$IFNDEF FPC}, Windows{$ENDIF},`

There seems to be a hardcoded rule in many places: When compiler is not FPC, then we are using Windows :-P

Then comes the following code:
```
procedure OpenSSLSetCurrentThreadID(id : PCRYPTO_THREADID); cdecl;
begin
  CRYPTO_THREADID_set_numeric(id, TOpenSSL_C_ULONG(GetCurrentThreadId));
end;
```

`GetCurrentThreadId` is a Windows call. Here is the correct way of using CurrentThreadId on multiplatforms:

```
<IdGlobal.pas>

function CurrentThreadId: TIdThreadID;
{$IFDEF USE_INLINE}inline;{$ENDIF}
begin
{$IFDEF DOTNET}
  ...
  Result := System.Threading.Thread.CurrentThread.ManagedThreadId;
  ...
{$ELSE}
  // TODO: is GetCurrentThreadId() available on Linux?
  Result := GetCurrentThreadID;
{$ENDIF}
end;
```

I hope we are getting one step closer to Linux with this discussion :-)

RoLex commented:
> Yes, but only if `MSWINDOWS` is defined by the compiler. However, it does also manually define `MSWINDOWS` prior to Delphi 6, so is it possible that `VCL_6_OR_ABOVE` is somehow not being defined correctly for your Delphi compiler?

When I create a new project in my Delphi setup, without using Indy or IndyOpenSSL, and setting target platform to Linux, I get this result:

```
{$IFDEF VCL_6_OR_ABOVE}
  // not defined
{$ENDIF}
```

EDIT: So I have to include `{$I IdCompilerDefines.inc}` at the very beginning of my project code? It works when I do that. So VCL_6_OR_ABOVE was my mistake! `GetCurrentThreadId` is the next issue..

RoLex commented:
```
  {$IFNDEF VCL_6_OR_ABOVE}
    {$DEFINE MSWINDOWS} // is defined
  {$ENDIF}
  {$IFDEF MSWINDOWS}
    {$DEFINE WINDOWS} // is defined
  {$ENDIF}

  {$I IdCompilerDefines.inc} // IndySockets/Indy/blob/Indy-10.7/Lib/Core/IdCompilerDefines.inc

  {$IFNDEF VCL_6_OR_ABOVE}
    {$DEFINE MSWINDOWS} // not defined
  {$ENDIF}
  {$IFDEF MSWINDOWS}
    {$DEFINE WINDOWS} // is defined anyway
  {$ENDIF}
```

MSWINDOWS is defined anyway, and therefore WINDOWS is.

My includes go in following order: Indy/Core (branch: Indy-10.7), Indy/System (branch: Indy-10.7), Indy/Protocols (branch: Indy-10.7), IndyTLSOpenSSL (branch: master by MWASoftware)

This is driving me crazy! :-D

RoLex commented:
This is ridiculous..

<img width="1109" height="735" alt="IdSSLOpenSSL" src="https://github.com/user-attachments/assets/fb7aede0-050f-4572-968d-166601edcd40" />


Something is playing evil game with me. Again, how I set up Indy + OpenSSL:

Build and install as documented: IndySockets/Indy/tree/Indy-10.7
Build and install: MWASoftware/IndyTLS-OpenSSL/tree/master

Includes in Delphi:

```
<DCC_UnitSearchPath>
C:\Users\123\Programming\Delphi\Indy\Lib\Core;
C:\Users\123\Programming\Delphi\Indy\Lib\System;
C:\Users\123\Programming\Delphi\Indy\Lib\Protocols;
C:\Users\123\Programming\Delphi\MWASoftware_IndyTLS-OpenSSL\src;
C:\Users\123\Programming\Delphi\MWASoftware_IndyTLS-OpenSSL\src\opensslHdrs;
$(DCC_UnitSearchPath)
</DCC_UnitSearchPath>
```

MWASoftware committed:
Added AllTests group for Delphi

MWASoftware committed:
Fix Issue with incorrect condition ref to Windows unit in crypto#

MWASoftware committed:
Compiler defines should no longer define WINDOWS when compling under Delphi on POSIX platforms When VCL_6_OR_ABOVE is not defined. Possibly temp fix

MWASoftware committed:
Correct Delphi pkacage name dependency

Commit:
Change Lazarus package name to IndyTLSOpenSSL for consistency with Delphi version

Commit:
Correct Delphi Test Programs Package Dependency for OpenSSL

MWASoftware commented:
RoLex, your comments are much appreciated. I am able test with Lazarus/fpc on both Linux and Windows and with Delphi on Windows, but I don't have a Delphi/Linux combination and your feedback can help remove any config bugs for this platform.
I have uploaded a new set of commits to my [IndyTLS-OpenSSL](https://github.com/MWASoftware/IndyTLS-OpenSSL) repository. These should address:

1. {$IFNDEF FPC}Windows... in IdOpenSSLHeaders_crypto.pas
The header templates were from multiple contributors and this one just slipped through the net given that there is no error on any of my test platforms. I have changed the conditional to {$IFDEF WINDOWS} Windows. However, there may be an additional change needed for Delphi/Linux. the problematic function is "GetCurrentThreadID". For FPC (all versions) this is a System function. For Delphi/Windows, its in "Windows". The question is where is the declaration for Delphi/Linux? I need your help here to confirm that the change works, and if not where is this function declared.

2. Consistent package naming for Delphi and Lazarus: IndyTLSOpenSSL (lazarus) IndyTLSOpenSSLnnn (Delphi).
3. IdCompilerDefines.inc issue with spurious definition of WINDOWS.
If have put in an extra test for Posix platforms ie.
```pascal
  {$IFNDEF VCL_6_OR_ABOVE}
    {$IFNDEF POSIX}
    {$DEFINE MSWINDOWS}
    {$ENDIF}
  {$ENDIF}
```
This should fix the issue for now. However, Remy needs to investigate why VCL_6_OR_ABOVE was not defined. Perhaps Delphi/ Linux uses a VCL version number not recognised by Indy.

RoLex commented:
> However, there may be an additional change needed for Delphi/Linux. the problematic function is "GetCurrentThreadID". For FPC (all versions) this is a System function. For Delphi/Windows, its in "Windows". The question is where is the declaration for Delphi/Linux? I need your help here to confirm that the change works, and if not where is this function declared.

Thank you for your work so far!

As I wrote in one of my previous posts, you can either include `IdGlobal` and use `CurrentThreadId`, or include `Posix.Pthread` if platform is `POSIX` - this solution worked for me.

@rlebeau 

This todo comment can be removed, because `IdGlobal.pas` includes `Posix.Pthread` and compiles well with Linux: https://github.com/IndySockets/Indy/blob/Indy-10.7/Lib/System/IdGlobal.pas#L5884

@MWASoftware 

```
[DCC Error] IdSSLOpenSSLAPI.pas(483): E2003 Undeclared identifier: 'DirListDelimiter'
[DCC Error] IdSSLOpenSSLAPI.pas(502): E2003 Undeclared identifier: 'DefaultLibVersions'
```

Solution: `IdSSLOpenSSLAPI.pas` needs `{$I IdCompilerDefines.inc}` to set compiler flags correctly.

```
[DCC Error] IdSSLOpenSSLAPI.pas(176): E2029 $IFEND expected but $ENDIF found
[DCC Error] IdSSLOpenSSLAPI.pas(218): E2029 $IFEND expected but $ENDIF found
```

Solution: As suggested by compiler.

```
[DCC Error] IdSSLOpenSSLAPI.pas(542): E2010 Incompatible types: 'PWideChar' and 'AnsiString'
[DCC Error] IdSSLOpenSSLAPI.pas(548): E2010 Incompatible types: 'PWideChar' and 'AnsiString'
```

Solution: As compiler suggests: `{$IFDEF UNIX}PChar(AProcName){$ELSE}PAnsiChar(AProcName){$ENDIF}`

Now everything compiles! Next step will be testing my code in production, I will get back with results :)

Thank you both very much!

<img width="972" height="1013" alt="IdSSLOpenSSLAPI_1" src="https://github.com/user-attachments/assets/3859b790-e10e-4335-b1ef-1f40d05eae4f" />
<img width="723" height="299" alt="IdSSLOpenSSLAPI_2" src="https://github.com/user-attachments/assets/05590930-7745-4dc7-a81e-a93bff7f3c0a" />

MWASoftware committed:
OpenSSLAPI fixes for Delphi on Linux

MWASoftware commented:
Thanks for your feedback.

My aim is for the OpenSSL headers to be independent of Indy and I have a separate repo PascalAPI4OpenSSL where they are maintained. The idea is that they are more widely applicable than just Indy and should be available for anyone who needs Pascal bindings for OpenSSL. Therefore the fixes have to be independent of IdCompilerDefines.

The fix for ..crypto.pas I propose is

```pascal
 {$IFNDEF FPC}
     {$IFDEF WINDOWS}
       ,Windows
     {$ELSE}
       ,Posix.Pthread
     {$ENDIF}
   {$ENDIF};
```
In OpensslAPI, the problem seems to be that Delphi does not define UNIX while fpc does not define POSIX. I thus propose to simply add a {$DEFINE UNIX} if POSIX is defined. Along with fixing the ifend issue (Delphi for Linux seems fussier here than Delphi for Windows) should wrap it up. Will add an update when the tested fix is available.

RoLex commented:
Correcting myself:

```
function LoadLibCryptoFunction(const AProcName: AnsiString): Pointer;
function LoadLibSSLFunction(const AProcName: AnsiString): Pointer;
```

Needs to be changed to:

```
function LoadLibCryptoFunction(const AProcName: PChar): Pointer;
function LoadLibSSLFunction(const AProcName: PChar): Pointer;
```

Without conversion of GetProcAddress parameter, as GetProcAddress expects a pointer.

Else the library calling is not working, atleast on Linux.

MWASoftware committed:
Further Updates for Delphi on Linux compatibility

MWASoftware committed:
LoadLib functions now use a PChar type param for Delphi on Linux compat

MWASoftware commented:
I've changed these functions as you suggested and tested on Linux (lazarus/fpc) and Windows (Lazarus/fpc and Delphi). All seems to work. Please confirm for Delphi on Linux. All changes uploaded to my github repo.
> Correcting myself:
> 
> ```
> function LoadLibCryptoFunction(const AProcName: AnsiString): Pointer;
> function LoadLibSSLFunction(const AProcName: AnsiString): Pointer;
> ```
> 
> Needs to be changed to:
> 
> ```
> function LoadLibCryptoFunction(const AProcName: PChar): Pointer;
> function LoadLibSSLFunction(const AProcName: PChar): Pointer;
> ```
> 
> Without conversion of GetProcAddress parameter, as GetProcAddress expects a pointer.
> 
> Else the library calling is not working, atleast on Linux.

RoLex commented:
Thank you!

There are further issues with Ansi vs Unicode strings when passing parmeters in procedures. We take this function as example:

```
<IdSSLOpenSSLX509.pas>

function IndySSL_CTX_use_certificate_chain_file(ctx :PSSL_CTX; const AFileName: String) : TIdC_INT;
{$IFDEF USE_INLINE} inline; {$ENDIF}
begin
  Result := SSL_CTX_use_certificate_chain_file(ctx, PAnsiChar(AFileName));
end;
```

`AFileName` is Unicode string in modern Delphi, while `PAnsiChar` is a pointer to Ansi string.

As Delphi compiler warns, there is a suspicious cast from Unicode to Ansi, so these functions are not working in production because strings are passed wrong way. I guess that FPC handles strings correctly automatically or always use Ansi/Raw independently - never worked with FPC, while in Delphi the Ansi vs Unicode compatibility headache has been known since forever :-P Also this in not Linux specific, meaning Windows is affected too.

This is how I fixed the issue: `PAnsiChar(AnsiString(AFileName))`

Indy is using `STRING_IS_UNICODE` and `STRING_IS_ANSI` ifdefs for backward compatibility with old non-Unicode Delphi compilers.

Functions that I've corrected so far:

```
IndyX509_STORE_load_locations
IndySSL_CTX_use_certificate_chain_file
IndySSL_CTX_use_certificate_file
IndySSL_CTX_use_PrivateKey_file
IndySSL_load_client_CA_file
```

Also as I suggested you earlier, `LoadLibCryptoFunction` and `LoadLibSSLFunction` is possibly wrong - not 100% sure, because if we pass PChar in parameters, the memory pointer might get lost when calling next function `GetProcAddress` and exiting current function.

I've changed to following instead:

```
function LoadLibCryptoFunction(const AProcName: String): Pointer;
begin
  Result := GetProcAddress(TOpenSSLDynamicLibProvider.FOpenSSLDDL.GetLibCryptoHandle, PChar(AProcName));
end;
```

Note that `GetProcAddress` is platform-independent in newer Delphi as it uses marshaller.

So far my code is working in production with these changes! :-D That is loading OpenSSL libraries, creating IO handler, loading certificates and accepting TLS connections on server.

MWASoftware commented:
> There are further issues with Ansi vs Unicode strings when passing parmeters in procedures. We take this function as example:
> 
> ```
> <IdSSLOpenSSLX509.pas>
> 
> function IndySSL_CTX_use_certificate_chain_file(ctx :PSSL_CTX; const AFileName: String) : TIdC_INT;
> {$IFDEF USE_INLINE} inline; {$ENDIF}
> begin
>   Result := SSL_CTX_use_certificate_chain_file(ctx, PAnsiChar(AFileName));
> end;
> ```
>As Delphi compiler warns, there is a suspicious cast from Unicode to Ansi, so these functions are not working in >production because strings are passed wrong way. I guess that FPC handles strings correctly automatically or always use >Ansi/Raw independently - never worked with FPC, while in Delphi the Ansi vs Unicode compatibility headache has been >known since forever :-P Also this in not Linux specific, meaning Windows is affected too.

>This is how I fixed the issue: PAnsiChar(AnsiString(AFileName))

>Indy is using STRING_IS_UNICODE and STRING_IS_ANSI ifdefs for backward compatibility with old non-Unicode Delphi >compilers.

Firstly, FPC is by default a UTF8 environment. Type string = ansistring and that is true for FPC under both Linux and Windows. On the other hand, Delphi is by default a UCS-2 environment. Type string = widestring. That I know is true for Windows and you appear to be confirming that that is also true for Delphi on Linux.

Secondly, all OpenSSL API calls use PAnsiChar under both Linux and Windows.

The code for IndySSL_CTX_use_certificate_chain_file predates my involvement (perhaps Remy can shed some light on the history here). The code

```pascal
function IndySSL_CTX_use_certificate_chain_file(ctx :PSSL_CTX;
  const AFileName: String) : TIdC_INT;
{$IFDEF USE_INLINE} inline; {$ENDIF}
begin
  Result := SSL_CTX_use_certificate_chain_file(ctx, PAnsiChar(AFileName));
end;
```

is only called when STRING_IS_UNICODE is not defined. When STRING_IS_UNICODE is defined, a rather convoluted variant is used which loads the file into memory and then passes it to OpenSSL. I have preferred to leave this code alone given that it seems to work.

I have not tried out an explict conversion of a Widestring to an AnsiString and then calling the OpenSSL API call (as you have suggested). I would expect it to work - but then why was this complicated approach done in the first place? I would like to know the history here before changing anything. 

For you to have issues here, it looks like STRING_IS_UNICODE is not defined for Delphi under Linux. If true, that is probably an error and is probably related to the earlier problem with "WINDOWS" being on a POSIX platform. Note that STRING_IS_UNICODE is defined consequential on VCL_2009_OR_ABOVE being defined.

I've noticed that the IdCompilerDefines.inc shipped with my fork of IndyTLS-OpenSSL does not include a line for VER350 (Alexandria) - the most recent version. That needs correcting and may be the reason why you are getting these problems.

MWASoftware commented:
> Also as I suggested you earlier, `LoadLibCryptoFunction` and `LoadLibSSLFunction` is possibly wrong - not 100% sure, because if we pass PChar in parameters, the memory pointer might get lost when calling next function `GetProcAddress` and exiting current function.
> 
> I've changed to following instead:
> 
> ```
> function LoadLibCryptoFunction(const AProcName: String): Pointer;
> begin
>   Result := GetProcAddress(TOpenSSLDynamicLibProvider.FOpenSSLDDL.GetLibCryptoHandle, PChar(AProcName));
> end;
> ```
Clearly this will work- although I am not sure whether there is any issue with the first approach. But what you suggest is probably safer and will move to what you suggest.

MWASoftware committed:
Merged in Upstream

MWASoftware committed:
Updated CompilerDefines to recognise VER350 (Delphi Alexandria)

MWASoftware commented:
I have uploaded an updated version of the package which includes an IdCompilerDefines.inc (taken from upstream) that should now recognise VER350.

RoLex commented:
`UTC_Time_Decode` in `IdSSLOpenSSLUtils.pas` is not inline, while `UTCTime2DateTime` that uses `UTC_Time_Decode` is inline - got compilation error there.

Today I will be testing your latest changes and get back with results.

rlebeau commented:
> This should fix the issue for now. However, Remy needs to investigate why VCL_6_OR_ABOVE was not defined. Perhaps Delphi/ Linux uses a VCL version number not recognised by Indy.

Delphi doesn't use VCL at all for cross-platform development.

In any case, `VCL_6_OR_ABOVE` should be defined in all versions of Delphi from v6 onward. If it is not, then something is wrong.

Also, FreePascal defines `WINDOWS`, but Delphi defines `MSWINDOWS`, so Indy defines `WINDOWS` for Delphi if `MSWINDOWS` is defined. Indy *should* be defining `MSWINDOWS` only for Delphi v5 and earlier, as Delphi started defining `MSWINDOWS` in v6 (when CLX was first introduced).

If `WINDOWS` is getting defined on Linux, then something is seriously wrong.

MWASoftware commented:
> In any case, `VCL_6_OR_ABOVE` should be defined in all versions of Delphi from v6 onward. If it is not, then something is wrong.

I think I have solved the problem. My fork of IndyTLS-OpenSSL seems to have failed to pick up the updated version of IdCompilerDefines.inc and thus was not recognising VER350 (Alexandria) - hence VCL_6_OR_ABOVE. Now corrected. It's a pity that Delphi does not have the equivalent of FPC_FULLVERSION and related defines.

RoLex commented:
> I think I have solved the problem. My fork of IndyTLS-OpenSSL seems to have failed to pick up the updated version of IdCompilerDefines.inc and thus was not recognising VER350 (Alexandria) - hence VCL_6_OR_ABOVE. Now corrected.

That is right, everything works as expected on Delphi 12, both for Windows and Linux! There are some unidentified variables and deprecated functions usage left, but we can take these later. Thank you very much for your work MWASoftware!

MWASoftware commented:
> > I think I have solved the problem. My fork of IndyTLS-OpenSSL seems to have failed to pick up the updated version of IdCompilerDefines.inc and thus was not recognising VER350 (Alexandria) - hence VCL_6_OR_ABOVE. Now corrected.
> 
> That is right, everything works as expected on Delphi 12, both for Windows and Linux! There are some unidentified variables and deprecated functions usage left, but we can take these later. Thank you very much for your work MWASoftware!

And thank you for helping to ensure that all is working on Delphi for Windows and Linux. An old version of IdCompilerDefines was the main issue, but the function signature of GetProcAddress and the location of GetCurrentThreadId where also important issues to have resolved.

How certificates are loaded in a unicode environment also needs to be addressed in a code tidy-up sense.

Otherwise, the package does seem good to go, and for many users is just a drop in replacement for Indy 6.

RoLex commented:
Not tested your code on Win64 before though, and here we go, there is a runtime exception on loading certificates:

<img width="1124" height="1030" alt="EIdOSSLLoadingCertError_x64" src="https://github.com/user-attachments/assets/f020b627-a0fd-4388-9f79-805a21d0f288" />

Win32 and Linux64 are fine! Using dynamically loaded `libcrypto-3-x64.dll` and `libssl-3-x64.dll`.

DLLs are compiled from `OpenSSL 3.4.1 11 Feb 2025`.

MWASoftware commented:
> Not tested your code on Win64 before though, and here we go, there is a runtime exception on loading certificates:
>
Interesting. Looks like an issue with the Wincrypt dll name. I am in the process of upgrading my copy of Delphi CE to Delphi 12 and should then be able ctx check this out.

MWASoftware commented:
> > Not tested your code on Win64 before though, and here we go, there is a runtime exception on loading certificates:
> >
> Interesting. Looks like an issue with the Wincrypt dll name. I am in the process of upgrading my copy of Delphi CE to Delphi 12 and should then be able ctx check this out.
> 
> 
Cancel the first response. I had already tested with 64 bit FPC and there was no problem with wincrypt. 

It looks like the problem is with loading your certificate. Are you using the original version of the Indy certificate loading functions, or your modified version coercing a unicode string to an ansistring?

RoLex commented:
Everything is original, with latest updates.

`mIOSSL.SSLOptions.CertFile := LCertPath` is all I do to set cert file, and same with key file option.

This works on Win32 with same cert files and 32-bit libs, works on Linux64 with same certs and preinstalled system libs, but not on Win64 with same certs and 64-bit libs.

Libs can be downloaded here:

https://wiki.overbyte.eu/arch/openssl-3.4.1-win64.zip
https://wiki.overbyte.eu/arch/openssl-3.4.1-win32.zip

rlebeau commented:
> It's a pity that Delphi does not have the equivalent of FPC_FULLVERSION and related defines.

Not as extensive as FPC does, but modern Delphi versions do have constants in the `System` unit which can be used with `{$IF}` (ie, `CompilerVersion`, `RTLVersion`, `RTLVersionXXX`, etc).  Indy doesn't use `{$IF}` with Delphi at this time due to backwards compatibility with Delphi 5 (`{$IF}` was introduced in Delphi 6), but when support for pre-Unicode compilers is dropped then I already have code written that will switch to `{$IF}` and stop using the `VERXXX` conditionals.

MWASoftware committed:
Updated for Delphi 10

MWASoftware committed:
Updated Delphi Project Files for Windows 11 and Delphi 12

MWASoftware committed:
Restore Win64 Delphi configuration

MWASoftware commented:
> Everything is original, with latest updates.
> 
> `mIOSSL.SSLOptions.CertFile := LCertPath` is all I do to set cert file, and same with key file option.
> 
> This works on Win32 with same cert files and 32-bit libs, works on Linux64 with same certs and preinstalled system libs, but not on Win64 with same certs and 64-bit libs.

I think I've found the problem. On a clean install of Windows-11 and Delphi 12 (CE), I found a problem with the order in which the package searches for the OpenSSL libraries. Instead of picking up the expected versions, it found a libcrypt.dll in Windows\System32. This did not seem to be compatible.

I have updated the search order in the updated version so that a version without a suffix is only picked up if there is no version with a suffix e.g. libcrypt-3-x64.dll,

RoLex commented:
> I have updated the search order in the updated version so that a version without a suffix is only picked up if there is no version with a suffix e.g. libcrypt-3-x64.dll

Hm, aren't `libcrypt` and `libcrypto` different libs? OpenSSL does not provide `libcrypt`, but `libcrypto`.

MWASoftware commented:
You are correct. I missed the 'o' out in my comment. The  correct name is used by the software.

MWASoftware committed:
The Find OpenSSL Lib algorithm no longer looks for a library without any suffix, unless an empty suffix is included in the list of suffixes. For Linux, the default empty suffix is the first in the list (e.g. looks for libcrtypo.so first. For Windows, libcrypto.dll is the last in the list.

MWASoftware commented:
> > It's a pity that Delphi does not have the equivalent of FPC_FULLVERSION and related defines.
> 
> Not as extensive as FPC does, but modern Delphi versions do have constants in the `System` unit which can be used with `{$IF}` (ie, `CompilerVersion`, `RTLVersion`, `RTLVersionXXX`, etc). Indy doesn't use `{$IF}` with Delphi at this time due to backwards compatibility with Delphi 5 (`{$IF}` was introduced in Delphi 6), but when support for pre-Unicode compilers is dropped then I already have code written that will switch to `{$IF}` and stop using the `VERXXX` conditionals.

Well, coming from a Safety Critical background where long term software support is essential, I'm impressed by your commitment to supporting a 25 year old product. I can just about remember Delphi 5 (I probably have the installation CD somewhere), but I upgraded to Delphi 6 almost as soon as it came out in 2001.

This is probably a good time to drop support for a compiler that does not support the {$if declared... and {$if defined,,,} syntax. 

Checking my proposed IndyTLS-OpenSSL package, I have made some use of {$if declared for type definitions. It's a very neat way of dealing with type definitions across multiple platforms and compilers. I could regress to just use of $IFDEF, but it would mean introducing all that clunky stuff in IdCompilerDefines.inc plus the maintenance headache implied.

Peter will probably have an even bigger problem with TaurusTLS if you keep that restraint. He took a snapshot of my OpenSSL headers early on in their development, and that iteration made very extensive use of both {$if defined and {$if declared conditionals. I've checked and they are still there in his repo. In later versions of the headers, I introduced the "Just in time" approach to function loading and that resulted in simpler code that did not need or use  {$if. He will need to reintegrate the current versions of the OpenSSL headers. A quick search of TaurusTLS.pas also revealed extensive use of this syntax.

MWASoftware commented:
> have updated the search order in the updated version so that a version without a suffix is only picked up if there is no version with a suffix e.g. libcrypt-3-x64.dll,

In more testing, I have found that with win64 in particular, it is very easy to pick up the libcrypto.ddl in Windows\system32 if you have not been careful to ensure that your version (e.g. libcrypto-3-x64.dll) is not properly installed. The result is the type of error @RoLex has been reporting. This is not great as the non-expert user will be given a misleading error report.

I am think of taking a more limited default approach to which filenames are permissible i.e. restricting those acceptable to libcrypto-1-x64.dll and libcrypto-3-x64.dll, whilst allowing the programmer to extend the list if they want to.

The problem has been that Windows Openssl dlls have had many different names over the years, starting with libeay32.dll. Maybe now is the time to be less permissive with the default dll names that are searched, and add a good readme on the issue.

rlebeau commented:
> Well, coming from a Safety Critical background where long term software support is essential, I'm impressed by your commitment to supporting a 25 year old product.

It's only because I know that there are still Indy users out there using such old versions.  But I'm also not against cutting them off at some point.  Have to move forward eventually.

> This is probably a good time to drop support for a compiler that does not support the {$if declared... and {$if defined,,,} syntax.

I'm aware. It's already implemented for Indy 11, where the minimum versions are moved up to Delphi 2009 and FPC 3.0.  Might be worth backporting some of the changes for 11 into 10.7. Not sure yet. I need to re-sync the Indy 11 code with the latest Indy 10 code.

> Peter will probably have an even bigger problem with TaurusTLS if you keep that restraint. He took a snapshot of my OpenSSL headers early on in their development, and that iteration made very extensive use of both {$if defined and {$if declared conditionals. I've checked and they are still there in his repo.

AFAIK, he's not targetting old Delphi versions.

RoLex commented:
> In more testing, I have found that with win64 in particular, it is very easy to pick up the libcrypto.ddl in Windows\system32 if you have not been careful to ensure that your version (e.g. libcrypto-3-x64.dll) is not properly installed. The result is the type of error @RoLex has been reporting. This is not great as the non-expert user will be given a misleading error report.

Is it possible to supply library name/path to the OpenSSL reader class? This was possible with PR#299, and the idea was good. I'm thinking first of all about people who will use my application. They will probably run on very old OS that does not support installing packages with new version of OpenSSL library. Now I'm talking about Linux with default package installer.

MWASoftware commented:
> Is it possible to supply library name/path to the OpenSSL reader class? This was possible with PR#299, and the idea was good. I'm thinking first of all about people who will use my application. They will probably run on very old OS that does not support installing packages with new version of OpenSSL library. Now I'm talking about Linux with default package installer.

Yes, there is an interface that allows control over the library loading. There is an example of use in the test programs, but I probably should add a README on its use. The interface is in IdSSLOpenSSLAPI.pas and is:

```pascal
type
  IOpenSSLDLL = interface(IOpenSSL)
  ['{1d6cd9e7-e656-4981-80d2-288b12a69306}']
    procedure SetOpenSSLPath(const Value: string);
    function GetSSLLibVersions: string;
    procedure SetSSLLibVersions(AValue: string);
    function GetLibCryptoHandle: TLibHandle;
    function GetLibSSLHandle: TLibHandle;
    function GetLibCryptoFilePath: string;
    function GetLibSSLFilePath: string;
    function GetFailedToLoadList: TStrings;
    function Load: Boolean;
    procedure Unload;
    function IsLoaded: boolean;
    property SSLLibVersions: string read GetSSLLibVersions write SetSSLLibVersions;
end;

function GetIOpenSSLDDL: IOpenSSLDLL;
```
The interface allows you to specify the directory to search for the dll/so and the list of possible suffices for the library name. Reviewing the interface, it does not allow you to override the base name for the dll/so e,g, libcrypto, libssl. Perhaps that should be included. The default SSLpath is empty i.e. use the system default search path, and the default suffices list is (list in preference order):

```pascal
  DefaultLibVersions = ':.3:.1.1:.1.0.2:.1.0.0:.0.9.9:.0.9.8:.0.9.7:.0.9.6'; //Linux

  DefaultLibVersions = '-3-x64;-1-x64;'; //Windows 64-bit
  DefaultLibVersions = '-3;-1;'; //Windows 32-bit
```
Note the empty list elements at the start of the Linux list and end of the Windows lists. I am thinking of removing empty list element from the Windows lists but not the Linux lists.

Generally, Linux has fewer issues here than Windows. Most distros use the standard libnames from the OpenSSL package and so you get e.g. libssl.so.3 and a symlink from libssl.so to libssl.so.3. If you want to use a non-standard build, its often easier to set the LD_LIBRARY_PATH environment variable than it is to use the IOpenSSLDLL interface - the choice is yours.

MWASoftware commented:
> > Peter will probably have an even bigger problem with TaurusTLS if you keep that restraint. He took a snapshot of my OpenSSL headers early on in their development, and that iteration made very extensive use of both {$if defined and {$if declared conditionals. I've checked and they are still there in his repo.
> 
> AFAIK, he's not targetting old Delphi versions.

I suppose this begs the question as to what your intentions are with the IndyTLS-OpenSSL package. Is it also constrained to support Delphi 5, or is it allows to use{$if ...?

MWASoftware committed:
Extension of IOpenSSLDLL interface plus documentation

MWASoftware commented:
> Yes, there is an interface that allows control over the library loading. There is an example of use in the test programs, but I probably should add a README on its use.

I've added READMe.OpenSSL. Most of the text already existed, so quickly done. Also extended the IOpenSSLDLL interface to allow for lib base name changes.

MWASoftware committed:
Edits to README.OpenSSL

rlebeau commented:
> I suppose this begs the question as to what your intentions are with the IndyTLS-OpenSSL package.

My plan was for:

- IndyTLS-OpenSSL 1.0 to just be a straight extraction as-is of the current OpenSSL code from Indy's Protocols package, thus providing backwards compatibility for all compilers that Indy supports. Just add the new package reference to existing projects and existing code will just work as-is.

- Then update IndyTLS-OpenSSL to 2.0 with all new OpenSSL code, which can also include dropping old compilers.

But before that could happen, JP created TaurusTLS, so IndyTLS-OpenSSL might not even get a 2.0 now (1.0 for sure, that's a done deal for Indy 10.7).  I'm not sure yet, I'm still open to the possibility of a 2.0 if it makes sense.  Or maybe merge TaurusTLS and IndTLS-OpenSSL 2.0 together, who knows.

MWASoftware commented:
But do you expect TaurusTLS or IndyTLS v2 to still support Delphi 5?

> > I suppose this begs the question as to what your intentions are with the IndyTLS-OpenSSL package.
> 
> My plan was for:
> 
> - IndyTLS-OpenSSL 1.0 to just be a straight extraction as-is of the current OpenSSL code from Indy's Protocols package, thus providing backwards compatibility for all compilers that Indy supports. Just add the new package reference to existing projects and existing code will just work as-is.
> 
> - Then update IndyTLS-OpenSSL to 2.0 with all new OpenSSL code, which can also include dropping old compilers.
> 
> But before that could happen, JP created TaurusTLS, so IndyTLS-OpenSSL might not even get a 2.0 now (1.0 for sure, that's a done deal for Indy 10.7).  I'm not sure yet, I'm still open to the possibility of a 2.0 if it makes sense.  Or maybe merge TaurusTLS and IndTLS-OpenSSL 2.0 together, who knows.

MWASoftware commented:
I would also argue that there needs to be more haste in terms of Indy supporting a current version of OpenSSL.

All my use of Indy is for REST type services over https, and I suspect that the majority of users want that as well.

Not being able to support OpenSSL 3 is a major deficiency in Indy 10.6. You need to get Indy 10.7 with OpenSSL 3 support out of the door ASAP.

RoLex commented:
I also have no idea what Indy project developers are waiting for. If the work on new OpenSSL began long time ago, there would not appear concurrents like Taurus or eSeGeCe. They did appear because people needed TLS 1.3 as the rest of the world is moving forward.

At the same time, I also understand that Remy does not have time needed to perform this large work on Indy - a free project, he said that many times before. But right now there is barely need of doing anything, the code has already been written!

Also Indy has the opportunity that noone else have - it is included with RAD studio by default.

My suggestion on milestone:

* Indy 10.6 - legacy SSL support, original code
* Indy 10.7 - alpha version with OpenSSL 3 support
* Indy 11 - stable release

But this needs to be done fast, so people get time to test it. By that time we are hoping to see Indy 11 in RAD studio 13 updates.

MWASoftware commented:
> I also have no idea what Indy project developers are waiting for. If the work on new OpenSSL began long time ago, there would not appear concurrents like Taurus or eSeGeCe. They did appear because people needed TLS 1.3 as the rest of the world is moving forward.
> 
> At the same time, I also understand that Remy does not have time needed to perform this large work on Indy - a free project, he said that many times before. But right now there is barely need of doing anything, the code has already been written!
> 
> Also Indy has the opportunity that noone else have - it is included with RAD studio by default.

Thank you for what you have said as well as helping to tease out some remaining multiplatform issues from my fork of the IndyTLS-OpenSSL package.

I can very much appreciate the problem that Remy has as the single maintainer of Indy. For the last 15 years, I have been maintaining the Lazarus fork of IBX and when you sign up to this, you sign up to dealing with user support, bug and fixes and, in my case, keeping up-to-date with the many changes that have taken place in the Firebird RDMBMS since it split from InterBase. It's a serious commitment.

On the other hand, in my professional (i.e. paid for) work, I have several times had to deploy the aphorism "don't let the perfect be the enemy of the good" - and at times like this. You do need to test software before release, and it's why I put a lot of emphasis on a properly constructed test plan that includes the capability for regression testing. But, until a product starts being tested in the field, you cannot get full confidence in its operation. This is why you have alpha and beta pre-releases. IHMO, Indy with OpenSSL 3.x support is more than ready for alpha or even beta release and has been so for some time.

It's now more than a year since Remy asked Peter Mugaas and myself to help add OpenSSL 3.x support. A strawman package was produced and then Peter and myself went down different paths using the same initial package. Peter produced an add-on to Indy 10 which required changing the unit and class names (TaurusTLS). I developed a fork of Indy 10 with OpenSSL 3.x support (Indy.ProposedUpdate) and kept this is up-to-date with Indy 10 commits. I also copied the same OpenSSL units to a fork of Remy's proposed IndyTLS-OpenSSL package for use with the putative Indy 10.7 (and which you have been testing). At the very least, Indy users had a choice of upgrade path.

In terms of the code, mine and the TaurusTLS software has diverged since then. I would point to the work I have done since the initial strawman on improving the OpenSSL library load strategies with the faster "Just in Time" approach. I have also done some code clean up in the meantime in respect of the Indy OpenSSL code, and put emphasis on ensuring that both Delphi and Lazarus are fully supported by the same codebase - plus many bug fixes

For Indy users, there are two alternatives available at present: use Indy 10 with a separate add-on and your change class and unit names, or fully replace the distributed Indy 10 with my Indy.Proposed update fork and avoid name changes. That's the user's choice.

However, it would by much better if there was an "offical" path that was properly distributed with the next update of Delphi and via the Lazarus Online Package Manager (OPM). I don't see why Indy 10.7 can't be made available now (label it an alpha if you wish - just get it out there). As to whether the OpenSSL support path is via TaurusTLS or IndyTLS-OpenSSL - well naturally I favour the latter. But if TaurusTLS was preferred then I would be happy to offer to fold in the work I have done on improvements to the OpenSSL headers, library loading and code clean up - along with comparing bug fixes, into TaurusTLS. The important thing is to get something out there.

Finally, I have one further vested interest to declare. I have an oauth2client package on github that is accumulating stars and depends on Indy. I would really like to make this available via the Lazarus OPM. Indy 10 is in the OPM and there is no point in distributing this package if it can then only work with obsolete OpenSSL libraries and provides at most TLS 1.2 (now often deprecated). What I want is to get Indy 10.7 and IndyTLS-OpenSSL distributed via the OPM.

RoLex commented:
Btw, Win64 now works with Delphi 12! =)

I guess it's time to get rid of last small warnings and hints from Delphi compiler:

```
[dcc64 Hint] IdOpenSSLHeaders_crypto.pas(1080): H2077 Value assigned to 'COMPAT_OPENSSL_init_crypto' never used
[dcc64 Warning] IdOpenSSLHeaders_ssl.pas(5193): W1035 Return value of function 'COMPAT_SSL_new_session_ticket' might be undefined
[dcc64 Warning] IdSSLOpenSSLExceptionHandlers.pas(76): W1000 Symbol 'StrPas' is deprecated: 'Moved to the AnsiStrings unit'
[dcc64 Warning] IdSSLOpenSSLExceptionHandlers.pas(76): W1057 Implicit string cast from 'AnsiString' to 'string'
[dcc64 Warning] IdSSLOpenSSLAPI.pas(414): W1057 Implicit string cast from 'AnsiChar' to 'string'
[dcc64 Warning] IdSSLOpenSSLAPI.pas(419): W1057 Implicit string cast from 'AnsiChar' to 'string'
[dcc64 Hint] IdSSLOpenSSLFIPS.pas(208): H2164 Variable 'LRet' is declared but never used in 'OpenSSLGetSHA1HashInst'
[dcc64 Warning] IdSSLOpenSSLFIPS.pas(250): W1035 Return value of function 'OpenSSLGetSHA224HashInst' might be undefined
[dcc64 Hint] IdSSLOpenSSLFIPS.pas(239): H2164 Variable 'LRet' is declared but never used in 'OpenSSLGetSHA224HashInst'
[dcc64 Warning] IdSSLOpenSSLFIPS.pas(279): W1035 Return value of function 'OpenSSLGetSHA256HashInst' might be undefined
[dcc64 Hint] IdSSLOpenSSLFIPS.pas(268): H2164 Variable 'LRet' is declared but never used in 'OpenSSLGetSHA256HashInst'
[dcc64 Hint] IdSSLOpenSSLFIPS.pas(297): H2164 Variable 'LRet' is declared but never used in 'OpenSSLGetSHA384HashInst'
[dcc64 Hint] IdSSLOpenSSLFIPS.pas(328): H2164 Variable 'LRet' is declared but never used in 'OpenSSLGetSHA512HashInst'
[dcc64 Hint] IdSSLOpenSSLSocket.pas(301): H2443 Inline function 'TCriticalSection.Enter' has not been expanded because unit 'System.SyncObjs' is not specified in USES list
[dcc64 Hint] IdSSLOpenSSLSocket.pas(312): H2443 Inline function 'TCriticalSection.Leave' has not been expanded because unit 'System.SyncObjs' is not specified in USES list
[dcc64 Hint] IdSSLOpenSSLSocket.pas(332): H2443 Inline function 'TCriticalSection.Enter' has not been expanded because unit 'System.SyncObjs' is not specified in USES list
[dcc64 Hint] IdSSLOpenSSLSocket.pas(378): H2443 Inline function 'TCriticalSection.Leave' has not been expanded because unit 'System.SyncObjs' is not specified in USES list
[dcc64 Hint] IdSSLOpenSSLSocket.pas(400): H2443 Inline function 'TCriticalSection.Enter' has not been expanded because unit 'System.SyncObjs' is not specified in USES list
[dcc64 Hint] IdSSLOpenSSLSocket.pas(427): H2443 Inline function 'TCriticalSection.Leave' has not been expanded because unit 'System.SyncObjs' is not specified in USES list
[dcc64 Warning] IdSSLOpenSSLX509.pas(235): W1036 Variable 'LBufPtr' might not have been initialized
```

rlebeau commented:
> But do you expect TaurusTLS or IndyTLS v2 to still support Delphi 5?

Certainly not.

> I would also argue that there needs to be more haste in terms of Indy supporting a current version of OpenSSL.

I'm aware of that.  This has been an ongoing problem for several years now.

> I also have no idea what Indy project developers are waiting for.

I'm the only project developer left.  Its not that I'm *waiting*, its that I haven't had *free time* to work on it.

> If the work on new OpenSSL began long time ago, there would not appear concurrents like Taurus or eSeGeCe. They did appear because people needed TLS 1.3 as the rest of the world is moving forward.

I agree.

> At the same time, I also understand that Remy does not have time needed to perform this large work on Indy - a free project, he said that many times before. But right now there is barely need of doing anything, the code has already been written!

It's not just *writing* the code, but also *reviewing* and *testing* the MASSIVE amount of changes that this update involves. And that I have not been able to focus on.  Hence why I'm having Indy 10.7 break out OpenSSL completely from the main library, so its isolated and easier for community members to work on going forward.

> Also Indy has the opportunity that noone else have - it is included with RAD studio by default.

That is not as big an opportunity as you think it is.

> we are hoping to see Indy 11 in RAD studio 13 updates.

Embarcadero usually bundles Indy updates only in major IDE releases.  Non-interface-breaking bug fixes *might* get into minor IDE updates, but not big version changes. Even if I were to publish Indy 10.7 soon, or even Indy 11 (which BTW is another large Indy update), Embarcadero would not be able to bundle it until the next major IDE release (14+).  So don't expect to see 10.7 or 11 appear in RAD Studio 13 updates.

> IHMO, Indy with OpenSSL 3.x support is more than ready for alpha or even beta release and has been so for some time.

It has been in various forms, but getting a final final final release ready has been a challenge for me.

> For Indy users, there are two alternatives available at present

There have been a couple of other 3rd party implementations, too.

> However, it would by much better if there was an "offical" path that was properly distributed with the next update of Delphi and via the Lazarus Online Package Manager (OPM).

That's what I've been trying to get to.

MWASoftware committed:
Missing reference to IdSSLOpenSSLAPI added to IdSSLOpenSSLFIPS. Needed to use MD5 functions

MWASoftware committed:
Ensure that UseSystemRootCertificateStore property only applies to Windows environments

MWASoftware committed:
TIdSecSocket.Readable now checks for socket errors

MWASoftware committed:
Put package to top of build order in AllTests project group

RoLex commented:
https://github.com/IndySockets/IndyTLS-OpenSSL/pull/22/commits/c34ca1e164a6bf08cc83791d8dec056d8de343eb#diff-b9aefaa16cc17306ea293b47b3ab51bec71111dc609f055bcefdd0f7680ce3a2R9

@MWASoftware Here you set Win64 by default, which makes it impossible to build and install your package due to missing Indy packages. This is because Indy packages are built and installed for Win32 by default.

MWASoftware commented:
Thanks. Will update. I'm checking through the install and test procedures on various platforms and will feed this in.
> https://github.com/IndySockets/IndyTLS-OpenSSL/pull/22/commits/c34ca1e164a6bf08cc83791d8dec056d8de343eb#diff-b9aefaa16cc17306ea293b47b3ab51bec71111dc609f055bcefdd0f7680ce3a2R9
> 
> @MWASoftware Here you set Win64 by default, which makes it impossible to build and install your package due to missing Indy packages. This is because Indy packages are built and installed for Win32 by default.

MWASoftware committed:
Change of Status

MWASoftware commented:
Its clear to me from @rlebeau's post that 10.7 is still some way off. I have also concluded that there needs to be a support path for Delphi users which will have 10.6 for the foreseeable future. @JPeterMugaas was probably correct in splitting off to create TaurusTLS.

On the other hand, I don't want to waste all the work that has gone into create my fork of IndyTLS-OpenSSL. I have thus created a new version with the names changed to ensure that it can be used alongside Indy 10.6.

You can find this at https://github.com/MWASoftware/IndySecOpenSSL

I have so far tested it with Delphi 12 and 10, both with the the supplied Indy and replaced with Indy 10.7. I have also tested it with Lazarus on both Linux and Windows. Mindful of @RoLex's comments on Win32. I have gone through the package project files and the whole thing seems much slicker. THere are also GUI examples. I have also taken the opportunity to write a comprehensive User Guide.

The User Guide is 68 pages long as I have added my own text (as appendicies) analysing TLS 1.3 and DTLS, and a tutoral on Security Concepts. The main body of the User Guide is intended to be an indepth guide to installation and use, please a reference section.

I have also ended support for my fork of IndyTLS-OpenSSL and Indy.ProposedUpdate with all future work focused in the new package.
